lisa/transformer.py

```diff
@@ -0,0 +1,176 @@
+    for transformer in transformers:
+        if transformer.name in references:
+            raise LisaException(
+                f"found duplicate transformer names: '{transformer.name}'"
+            )
+        reference = set()
+        reference.update(transformer.depend_on)
+        references[transformer.name] = reference
+        names.add(transformer.name)
+    while transformers:
+        is_processed: bool = False
+        for transformer in transformers:
+            reference = references[transformer.name]
+
+            if reference:
+                for reference_item in reference:
+                    if reference_item not in names:
+                        raise LisaException(
+                            f"transformer '{transformer.name}' "
+                            f"depends on non-existing transformer "
+                            f"'{reference_item}'"
+                        )
+            else:
+                sorted_transformers.append(transformer)
+                transformers.remove(transformer)
+                del references[transformer.name]
+                for other in references.values():
+                    if transformer.name in other:
+                        other.remove(transformer.name)
+                is_processed = True
+                break
+        if not is_processed and transformers:
+            raise LisaException(
+                f"found cycle dependented transformers: "
+                f"{[x.name for x in transformers]}"
+            )
+
+    return sorted_transformers
+
+
+def _run_transformers(
+    transfromers_data: List[Any],
+    variables: Dict[str, VariableEntry],
+    is_dry_run: bool = False,
+) -> Dict[str, VariableEntry]:
+    transformers_runbook: List[schema.Transformer] = []
+    for runbook_data in transfromers_data:
+        # get raw transformer runbook for replacing variables.
+        runbook = schema.Transformer.schema().load(runbook_data)  # type: ignore
+        transformers_runbook.append(runbook)
+
+    # resort the runbooks, and it's used in real run
+    transformers_runbook = _sort(transformers_runbook)
+
+    copied_variables: Dict[str, VariableEntry] = dict()
+    for value in variables.values():
+        copied_variables[value.name] = value.copy()
+
+    factory = subclasses.Factory[Transformer](Transformer)
+    for runbook in transformers_runbook:
+        # serialize to data for replacing variables
+        runbook_data = runbook.to_dict()
+
+        # replace to validate all variables exist
+        replace_variables(runbook_data, copied_variables)
+
+        # revert to runbook
+        runbook = schema.Transformer.schema().load(runbook_data)  # type: ignore
+
+        transformer = factory.create_by_runbook(runbook=runbook)
+        values = transformer.run(is_dry_run=is_dry_run)
+        merge_variables(copied_variables, values)
+
+    return copied_variables
+
+
+def run(runbook_builder: RunbookBuilder) -> None:
```

Comment:
Should we put these standlalone methods into a class called `TransformerGroup` or something similar?

Reply:
The Python supports patterns of both functional programming and OOP. They apply to different scenarios. The file of Python is considered as a module, so the functions are provided by module. It doesn't need to be wrapped in a class.